WalletWasabi/Services/TransactionProcessor.cs

```diff
@@ -0,0 +1,215 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using NBitcoin;
+using WalletWasabi.KeyManagement;
+using WalletWasabi.Models;
+
+namespace WalletWasabi.Services
+{
+	public class TransactionProcessor
+	{
+		public KeyManager KeyManager { get; }
+		public ConcurrentHashSet<uint256> TransactionHashes { get; }
+		public ConcurrentHashSet<SmartTransaction> TransactionCache { get; }
+		public ObservableConcurrentHashSet<SmartCoin> Coins { get; }
+		public Money DustThreshold { get; }
+		public event EventHandler<SmartCoin> CoinSpentOrSpenderConfirmed;
+		public event EventHandler<SmartCoin> CoinReceived; 
+
+		public TransactionProcessor(KeyManager keyManager, 
+			ConcurrentHashSet<uint256> transactionHashes, 
+			ObservableConcurrentHashSet<SmartCoin> coins,
+			Money dustThreshold
+		)
+		{
+			KeyManager = keyManager;
+			TransactionHashes = transactionHashes;
+			TransactionCache = new ConcurrentHashSet<SmartTransaction>();
+			Coins = coins;
+			DustThreshold = dustThreshold;
+		}
+
+		public bool Process(SmartTransaction tx)
+		{
+			uint256 txId = tx.GetHash();
+			var walletRelevant = false;
+
+			bool justUpdate = false;
+			if (tx.Confirmed)
+			{
+				TransactionHashes.TryRemove(txId); // If we have in mempool, remove.
+				if (!tx.Transaction.PossiblyP2WPKHInvolved())
+				{
+					return false; // We do not care about non-witness transactions for other than mempool cleanup.
+				}
+
+				bool isFoundTx = TransactionCache.Contains(tx); // If we have in cache, update height.
+				if (isFoundTx)
+				{
+					SmartTransaction foundTx = TransactionCache.FirstOrDefault(x => x == tx);
+					if (foundTx != default(SmartTransaction)) // Must check again, because it's a concurrent collection!
+					{
+						foundTx.SetHeight(tx.Height, tx.BlockHash, tx.BlockIndex);
+						walletRelevant = true;
+						justUpdate = true; // No need to check for double spend, we already processed this transaction, just update it.
+					}
+				}
+			}
+			else if (!tx.Transaction.PossiblyP2WPKHInvolved())
```

Comment:
Is it possible to check `if (!tx.Transaction.PossiblyP2WPKHInvolved())` first then `else if (tx.Confirmed)` second?
If yes then can we remove `if (!tx.Transaction.PossiblyP2WPKHInvolved())` after this line `TransactionHashes.TryRemove(txId);`?

Reply:
Yes, it is possible but this PR is not easy to review given it moves a big piece of code and it already contains a change. I don't want to make any other modification until this is merged.